gen3utils/validate.py

```diff
@@ -0,0 +1,280 @@
+            services:microservice name
+            manifest_versions: the versions block from manifest.json
+        Return:
+            microservice version
+    """
+    for manifest_version in manifest_versions:
+        if manifest_version == services:
+            # if this fails, make sure you check the service is in the manifest before
+            # calling manifest_version()! it should never happen
+            microservice_version = manifest_versions.get(services).split(":")[1]
+            if "_" in microservice_version:
+                logger.warning(services + " is on branch")
+            return microservice_version
+
+
+def versions_validation(manifest, versions_requirements):
+    """
+    Validates versions in cdis-manifest 
+
+    Arg:
+        manifest: manfiest.json
+        versions_requirements: the "versions" requirement under validation_config.yaml
+
+    Return:
+        ok(bool): whether the validation succeeded.
+    """
+    ok = True
+
+    for versions_requirement in versions_requirements:
+        requirement_key_list = list(versions_requirement.keys())
+        requirement_key = requirement_key_list[0]
+        if requirement_key in manifest["versions"] and "_" not in manifest_version(
+            manifest["versions"], requirement_key
+        ):
+            # If the first service set to * under validation_config versions, other services should be in the manifest
+            # The second condition is ignoring branch on sevice. WHICH IS NOT GOO. Added a warning in the log
+            if (
+                versions_requirement[requirement_key] == "*"
+                and manifest["global"].get("hostname") != "qa-mickey.planx-pla.net"
+            ):
+                for required_service in requirement_key_list[1:]:
+                    ok = (
+                        assert_and_log(
+                            required_service in manifest["versions"],
+                            required_service + " is missing in manifest.json",
+                        )
+                        and ok
+                    )
+
+            elif (
+                "min"
+                and "max" not in versions_requirement[requirement_key]
+                and versions_requirement[requirement_key]
+                <= manifest_version(manifest["versions"], requirement_key)
+            ):
+                # If the first service set to a specific version in validation_config, other services should matches the version requirements
+                ok = (
+                    version_requirement_validation(
+                        versions_requirement, requirement_key_list, manifest["versions"]
+                    )
+                    and ok
+                )
+            #                for required_service in requirement_key_list[1:]:
+            #                    ok = (
+            #                        assert_and_log(
+            #                            versions_requirement[required_service]
```

Comment:
why is this commented out?

Reply:
🤦‍♀ this should be removed. I added function version_requirement_validation to replace this part.